### DISCUSSION TITLE
Timeout for queries/mutation - backend does not response causing an infinite loading

### DISCUSSION THREAD

**Discussion by acuevas91** [enisdenjo/graphql-ws#467] [Q&A] [answered]
I have recently an issue regarding an infinite loading due to the app did not get any response for the backend so the spinner never ends.

My question is, despite the issue is in the backend, does anyone know if there is an option for adding a timeout for queries/mutations in case backend does not respond due to an issue?

Many thanks in advance!

**Comment by enisdenjo** [comment, marked as answer]
You can create a timeout yourself. Roughly:

```js
import { createClient } from 'graphql-ws';

const client = createClient({
  url: 'ws://localhost:4000/graphql',
});

let timeout;
const dispose = client.subscribe(
  {
    query: '{ slowQuery }',
  },
  {
    next: (value) => {
      // value received, clear timeout
      clearTimeout(timeout);
    },
    error: (err) => {},
    complete: () => {},
  },
);
// remove subscription after a timeout of 3 seconds
timeout = setTimeout(dispose, 3000);
```

**Reply by acuevas91** [reply]
Many thanks for your answer! I have one doubt regarding this issue, how could you return a custom error? As it is async and we cannot use try catch due to in sync...

I tried to add it in setTimeout handler but it did not work...

```
timeout = setTimeout(() => { 
   try {
     dispose();
   } catch(error) {
   // throw custom error
   }
 }, 3000);
```

**Reply by enisdenjo** [reply]
`dispose()` returns a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise), so you can simply use its primitives.
```js
timeout = setTimeout(() => dispose().catch((err) => console.error(err)), 3000);
```

Alternatively, you can pass in an async function to `setTimeout` and then use the try/catch block:
```diff
- timeout = setTimeout(() => { 
+ timeout = setTimeout(async () => { 
   try {
-     dispose();
+     await dispose();
   } catch(error) {
   // throw custom error
   }
 }, 3000);
```

**Reply by acuevas91** [reply]
Ohh I did not know it as the return type of `dispose()` is `() => void`, so typescript does not allow me to use the first option, so I am trying with the second one but it seems not to work initially.

Many thanks for your answers! I really appreciate them

**Reply by enisdenjo** [reply]
Oh my bad! I was referencing the [client `dispose()` function](https://github.com/enisdenjo/graphql-ws/blob/fbb763a662802a6a2584b0cbeb9cf1bde38158e0/src/client.ts#L975-L982). 😬 

The `dispose()` of a subscription is not async, [its just a regular ol' function](https://github.com/enisdenjo/graphql-ws/blob/fbb763a662802a6a2584b0cbeb9cf1bde38158e0/src/client.ts#L970-L973).

**Reply by acuevas91** [reply]
So we should emit in the sink error in the timeout right? I am figuring out how could I throw a custom timeout exception and catch this error at ngrx level

**Reply by enisdenjo** [reply]
Ok so you would like to report an error if the timeout happens? Not just stop the subscription? If so, you can simply wrap all this in a custom subscribe function and then dispose and report the error on timeout.

**Reply by acuevas91** [reply]
Yes, actually this is what I was trying out 😄 so, what you mean is to wrap `client.subscribe` function and if error, dispose and report the error in that subscription? am I right?

**Reply by enisdenjo** [reply]
Yep yep. Roughly:
```js
function subscribeWithTimeout(client, payload, sink) {
  let timeout;
  const dispose = client.subscribe(payload, {
    ...sink,
    next: (value) => {
      sink.next(value);
      if (timeout) {
        clearTimeout(timeout);
      }
    },
  });
  timeout = setTimeout(() => {
    dispose();
    sink.error(new Error('Subscribe has timed out.'));
  }, 3000);
  return () => {
    dispose();
    if (timeout) {
      clearTimeout(timeout);
    }
  };
}
```

**Reply by acuevas91** [reply]
Many thanks for all your support, that works!
